src/Appwrite/Platform/Workers/Messaging.php

```diff
@@ -179,13 +181,29 @@ private function processMessage(Database $dbForProject, Document $message): void
                         $data = match ($provider->getAttribute('type')) {
                             MESSAGE_TYPE_SMS => $this->buildSMSMessage($messageData, $provider),
                             MESSAGE_TYPE_PUSH => $this->buildPushMessage($messageData),
-                            MESSAGE_TYPE_EMAIL => $this->buildEmailMessage($messageData, $provider),
+                            MESSAGE_TYPE_EMAIL => $this->buildEmailMessage($dbForProject, $messageData, $provider),
                             default => throw new Exception(Exception::PROVIDER_INCORRECT_TYPE)
                         };
 
                         try {
-                            $adapter->send($data);
-                            $deliveredTotal += \count($batch);
+                            $response = new Response($provider->getAttribute('type'));
+                            $response->fromArray($adapter->send($data));
+
+                            $deliveredTotal += $response->getDeliveredTo();
+                            $details[] = $response->getDetails();
+                            foreach ($details as $detail) {
+                                if ($detail['status'] === 'failure') {
+                                    $deliveryErrors[] = `Failed sending to target {$detail['recipient']} with error: {$detail['error']}`;
+                                }
+
+                                // Deleting push targets when token has expired.
+                                if ($detail['error'] === 'Expired device token.') {
+                                    $target = $dbForProject->findOne('targets', [Query::equal('identifier', [$detail['recipient']])]);
```

Comment:
Shouldn't the targets already be available in `$recipients`? Do we need to make another database query? Perhaps `$recipients` should be a hash with the key being the identifier so you can lookup the Target faster here.

Reply:
No longer applicable after https://github.com/appwrite/appwrite/pull/7394